keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,674 @@
+
+Additionally, this KEP aims to improve the initialization and reinitialization of resource managers, such as the memory manager and CPU manager, so that they can adapt to change in node's configurations.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I must be able to increase the cluster resource capacity without adding a new node to the cluster.
+
+#### Story 2
+
+As a cluster admin, I must be able to increase the cluster resource capacity without need to restarting the kubelet.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+1. Node resource hot plugging is an opt-in feature, merging the
+   feature related changes won't impact existing workloads. Moreover, the feature
+   will be rolled out gradually, beginning with an alpha release for testing and
+   gathering feedback. This will be followed by beta and GA releases as the
+   feature matures and potential problems and improvements are addressed.
+2. Though the node resource is updated dynamically, the dynamic data is fetched from cAdvisor and its well integrated with kubelet.
+3. Resource manager are updated to adapt to the dynamic node reconfigurations, Enough tests should be added to make sure its not affecting the existing functionalities.
+
+## Design Details
+
+
+Below diagram is shows the interaction between kubelet and cAdvisor.
+
+
+```mermaid
+sequenceDiagram
+    participant node
+    participant kubelet
+    participant cAdvisor-cache
+    participant machine-info
+    kubelet->>cAdvisor-cache: Poll
+    cAdvisor-cache->>machine-info: fetch
+    machine-info->>cAdvisor-cache: update
+    cAdvisor-cache->>kubelet: update
+    kubelet->>node: node status update
+    kubelet->>node: re-initialize resource managers
+```
+
+The interaction sequence is as follows
+1. Kubelet will be polling in interval to fetch the machine resource information from cAdvisor's cache, Which is currently updated every 5 minutes.
+2. Kubelet's cache will be updated with the latest machine resource information.
+3. Node status updater will update the node's status with the latest resource information.
+4. Kubelet will reinitialize the resource managers to keep them up to date with dynamic resource changes.
+
+Note: In case of increase in cluster resources, the scheduler will automatically schedule any pending pods.
+
+**Proposed Code changes**
+
+**Dynamic Node Scale Up logic**
+
+```go
+	if utilfeature.DefaultFeatureGate.Enabled(features.NodeResourceHotPlug) {
+		// Handle the node dynamic scale up
+		machineInfo, err := kl.cadvisor.MachineInfo()
+		if err != nil {
+			klog.ErrorS(err, "Error fetching machine info")
+		} else {
+			cachedMachineInfo, _ := kl.GetCachedMachineInfo()
+
+			if !reflect.DeepEqual(cachedMachineInfo, machineInfo) {
```

Comment:
we don't need to debate this here, but are there other fields of machine info that may change more frequently? like timestamp of gathering for instance? we should make sure we're not needlessly churning

Reply:
Yeah, I will revisit and check for frequently changing values apart from actual resize values and update.